cmd/spire-server/cli/agent/agent_test.go

```diff
@@ -0,0 +1,290 @@
+		name               string
+		args               []string
+		expectedReturnCode int
+		expectedStdout     string
+		expectedStderr     string
+		existentAgent      string
+		dsError            error
+	}{
+		{
+			name:               "success",
+			args:               []string{"-spiffeID", "spiffe://example.org/spire/agent/agent1"},
+			expectedReturnCode: 0,
+			existentAgent:      "spiffe://example.org/spire/agent/agent1",
+			expectedStdout:     "Found an attested agent given its SPIFFE ID\n\nSPIFFE ID         : spiffe://example.org/spire/agent/agent1",
+		},
+		{
+			name:               "spiffe id not found",
+			args:               []string{"-spiffeID", "spiffe://example.org/spire/agent/agent2"},
+			expectedReturnCode: 1,
+			existentAgent:      "spiffe://example.org/spire/agent/agent1",
+			expectedStderr:     "rpc error: code = NotFound desc = agent not found\n",
+		},
+		{
+			name:               "no spiffe id",
+			expectedReturnCode: 1,
+			existentAgent:      "spiffe://example.org/spire/agent/agent1",
+			expectedStderr:     "a SPIFFE ID is required\n",
+		},
+		{
+			name:               "no agents - not found",
+			expectedReturnCode: 1,
+			args:               []string{"-spiffeID", "spiffe://example.org/spire/agent/agent1"},
+			expectedStderr:     "rpc error: code = NotFound desc = agent not found\n",
+		},
+		{
+			name:               "datastore error",
+			args:               []string{"-spiffeID", "spiffe://example.org/spire/agent/agent1"},
+			expectedReturnCode: 1,
+			dsError:            errors.New("datastore error"),
+			expectedStderr:     "rpc error: code = Internal desc = failed to fetch agent: datastore error\n",
+		},
+	} {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			test := setupTest(t)
+			if tt.existentAgent != "" {
+				test.createAgent(t, tt.existentAgent)
+			}
+
+			test.ds.SetNextError(tt.dsError)
+			returnCode := test.showCmd.Run(tt.args)
+			require.Contains(t, test.testEnv.Stdout.(*bytes.Buffer).String(), tt.expectedStdout)
+			require.Equal(t, tt.expectedStderr, test.testEnv.Stderr.(*bytes.Buffer).String())
+			require.Equal(t, tt.expectedReturnCode, returnCode)
+		})
+	}
+}
+
+func setupTest(t *testing.T) *agentTest {
+	ds := fakedatastore.New(t)
+
+	service := agent.New(agent.Config{
+		ServerCA:    fakeserverca.New(t, td.String(), &fakeserverca.Options{}),
+		DataStore:   ds,
+		TrustDomain: td,
+		Clock:       clock.NewMock(t),
+		Catalog:     fakeservercatalog.New(),
+	})
+
+	log, _ := test.NewNullLogger()
+	log.Level = logrus.DebugLevel
+	registerFn := func(s *grpc.Server) {
+		agent.RegisterService(s, service)
+	}
+
+	testEnv := &common_cli.Env{
+		Stdin:  new(bytes.Buffer),
+		Stdout: new(bytes.Buffer),
+		Stderr: new(bytes.Buffer),
+	}
+
+	test := &agentTest{
+		ds:      ds,
+		testEnv: testEnv,
+	}
+
+	contextFn := func(ctx context.Context) context.Context {
+		ctx = rpccontext.WithLogger(ctx, log)
+		return ctx
+	}
+
+	conn, done := spiretest.NewAPIServer(t, registerFn, contextFn)
```

Comment:
maybe not as part of this PR, but it's probably worth knowing if there are any consumers of `NewAPIServer` which call done "early" (i.e. before the test has finished) in order to exercise some scenario where server is down. If not, it is a good candidate to simply leverage t.Cleanup directly inside `NewAPIServer`...